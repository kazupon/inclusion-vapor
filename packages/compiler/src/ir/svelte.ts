// SPDX-License-Identifier: MIT
// Author: kazuya kawaguchi (a.k.a. kazupon)

import { makeMap, isObject } from '@vue-vapor/shared'

import type { SourceLocation } from '@vue-vapor/compiler-dom'
import type {
  Element as SvelteElement,
  BaseNode as SvelteBaseNode,
  Comment as SveletComment,
  Text as SvelteText,
  MustacheTag as SvelteMustacheTag
} from 'svelte/types/compiler/interfaces'

export const isBuiltInDirective: ReturnType<typeof makeMap> = /*#__PURE__*/ makeMap(
  // TODO: add svelte built-in directives
  ''
  //'bind,cloak,else-if,else,for,html,if,model,on,once,pre,show,slot,text,memo',
)

const SVELTE_ELEMENT_TYPES = new Set([
  'InlineComponent',
  'SlotTemplate',
  'Title',
  'Slot',
  'Element',
  'Head',
  'Options',
  'Window',
  'Document',
  'Body'
])

export function isSvelteElement(node: unknown): node is SvelteElement {
  return isObject(node) && 'type' in node && SVELTE_ELEMENT_TYPES.has((node as SvelteBaseNode).type)
}

const SVELETE_MUSTACHE_TAG_TYPES = new Set(['MustacheTag', 'RawMustacheTag'])

export function isSvelteMustacheTag(node: unknown): node is SvelteMustacheTag {
  return (
    isObject(node) &&
    'type' in node &&
    SVELETE_MUSTACHE_TAG_TYPES.has((node as SvelteBaseNode).type)
  )
}

export function isSvelteText(node: unknown): node is SvelteText {
  return isObject(node) && 'type' in node && node.type === 'Text'
}

export function isSvelteComment(node: unknown): node is SveletComment {
  return isObject(node) && 'type' in node && node.type === 'Comment'
}

export function convertToSourceLocation(node: SvelteBaseNode, source: string): SourceLocation {
  return {
    start: {
      offset: node.start,
      line: -1,
      column: -1
    },
    end: {
      offset: node.end,
      line: -1,
      column: -1
    },
    source
  }
}

export type {
  Ast as SvelteAst,
  Element as SvelteElement,
  TemplateNode as SvelteTemplateNode,
  Text as SvelteText,
  MustacheTag as SvelteMustacheTag,
  Comment as SvelteComment,
  Attribute as SvelteAttribute,
  SpreadAttribute as SvelteSpreadAttribute,
  BaseDirective as SvelteBaseDirective,
  BaseNode as SvelteBaseNode
} from 'svelte/types/compiler/interfaces'
