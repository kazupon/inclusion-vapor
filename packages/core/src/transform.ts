// import generate from '@babel/generator'
import { print as generate } from 'code-red'
import {
  analyze,
  Scope,
  ScopeType,
  Variable,
  DefinitionType
} from '@typescript-eslint/scope-manager'
import {
  simpleTraverse,
  parse as parseTsEslint,
  AST_NODE_TYPES
} from '@typescript-eslint/typescript-estree'
import { MagicStringAST } from 'magic-string-ast'

import type { SvelteScript } from './compiler'
import type { File as BabelFile, Node as BabelNode } from '@babel/types'

type TSESLintNode = ReturnType<typeof parseTsEslint>

export function transformSvelteVapor(code: string): ReturnType<typeof generate> {
  return {
    code,
    map: {}
  }
}

export function transformSvelteScript(script: SvelteScript, code: string): string {
  const babelFileNode = script.content as unknown as BabelFile
  const jsAst = babelFileNode.program as unknown as TSESLintNode
  enableParentableNodes(jsAst)

  const scopeManager = analyze(jsAst, { sourceType: 'module' })
  const scope = getScope(scopeManager, jsAst)
  const refVariables = getCanBeVaporRefVariables(scope)
  const jsStr = rewriteToVaporRef(refVariables, code, babelFileNode)

  // NOTE: avoid Maximum call stack size exceeded error with `code-red` print
  disableParentableNodes(jsAst)

  return jsStr.toString()
}

function rewriteToVaporRef(
  variables: Variable[],
  source: string,
  fileNode: BabelFile
): MagicStringAST {
  const s = new MagicStringAST(source)
  const offset = fileNode.start!
  for (const variable of variables) {
    const def = variable.defs[0]
    let importRef = false
    if (
      def.node.type === AST_NODE_TYPES.VariableDeclarator &&
      def.node.init &&
      def.node.init.type === AST_NODE_TYPES.Literal
    ) {
      s.overwriteNode(
        normalizeBabelNodeOffset(def.node.init as unknown as BabelNode, offset),
        `ref(${def.node.init.raw})`
      )
      importRef = true
    }
    if (importRef) {
      s.prepend(`import { ref } from 'vue/vapor'\n`)
    }
    const refs = variable.references.filter(
      ref =>
        ref.identifier !== variable.identifiers[0] &&
        ref.resolved?.identifiers[0] === variable.identifiers[0]
    )
    for (const ref of refs) {
      s.overwriteNode(
        normalizeBabelNodeOffset(ref.identifier as unknown as BabelNode, offset),
        `${ref.identifier.name}.value`
      )
    }
  }
  return s
}

function normalizeBabelNodeOffset(node: BabelNode, offset: number): BabelNode {
  return {
    ...node,
    start: node.start! - offset,
    end: node.end! - offset
  }
}

function getCanBeVaporRefVariables(scope: Scope): Variable[] {
  return scope.variables.filter(variable => {
    if (variable.defs[0]) {
      const def = variable.defs[0]
      if (
        def.type === DefinitionType.Variable &&
        def.name.type === AST_NODE_TYPES.Identifier &&
        def.node.type === AST_NODE_TYPES.VariableDeclarator &&
        def.node.init &&
        def.node.init.type === AST_NODE_TYPES.Literal
      ) {
        return variable
      }
    }
  })
}

function enableParentableNodes(node: TSESLintNode) {
  simpleTraverse(node, {
    enter(node, parent) {
      if (parent) {
        node.parent = parent
      }
    }
  })
}

function disableParentableNodes(node: TSESLintNode) {
  simpleTraverse(node, {
    enter(node) {
      if (node.parent) {
        // @ts-expect-error -- NOTE
        delete node.parent
      }
    }
  })
}

function getScope(manager: ReturnType<typeof analyze>, node: TSESLintNode): Scope {
  const scope = manager.acquire(node, true)
  if (scope) {
    if (scope.type === ScopeType.functionExpressionName) {
      return scope.childScopes[0]
    }
    return scope
  }
  return manager.scopes[0]
}
