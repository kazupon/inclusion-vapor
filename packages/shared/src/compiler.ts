// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)

export * from './ast.ts'
export * from './htmlNesting.ts'
export * from './types.ts'
