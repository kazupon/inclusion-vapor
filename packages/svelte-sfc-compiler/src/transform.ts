// SPDX-License-Identifier: MIT
// Author: kazuya kawaguchi (a.k.a. kazupon)

// import generate from '@babel/generator'
// import { print as _generate } from 'code-red'
import { parse as parseBabel } from '@babel/parser'
import { walkImportDeclaration } from 'ast-kit'
import { analyze as analyzeAst, getExportVariables, getReferences } from 'inclusion-vapor-shared'
import { generateTransform, MagicStringAST } from 'magic-string-ast'

import type {
  Identifier as BabelIdentifier,
  ImportDeclaration as BabelImportDeclaration,
  LabeledStatement as BabelLabeledStatement,
  Node as BabelNode,
  Program as BabelProgram
} from '@babel/types'
import type { ImportBinding } from 'ast-kit'
import type { Scope, Variable } from 'inclusion-vapor-shared'
import type { SvelteScript } from 'svelte-vapor-template-compiler'

type GenerateMap = ReturnType<typeof MagicStringAST.prototype.generateMap>

export function transformSvelteVapor(code: string): { code: string; map: GenerateMap } {
  return {
    code,

    map: {} as unknown as GenerateMap
  }
}

/**
 * {@link transformSvelteScript} options
 */
export interface TransformSvelteScriptOptions {
  /**
   * Svelte Script AST
   * @default undefined
   */
  ast?: SvelteScript
  /**
   * Svelte script id. if you want to use source map, you should set this.
   * @default undefined
   */
  id?: string
  /**
   * Enable source map
   * @default false
   */
  sourcemap?: boolean
}

/**
 * Transform Svelte script to Vapor script
 *
 * @param {string} code - a string of svelte script
 * @param {TransformSvelteScriptOptions} options - {@link TransformSvelteScriptOptions | options}
 * @returns {string | ReturnType<typeof generateTransform>}
 */
// TODO: frientdly return type with infer from options
export function transformSvelteScript(
  code: string,
  options: TransformSvelteScriptOptions = {}
): string | { code: string; map: GenerateMap } {
  const babelFileNode = options.ast
    ? options.ast.content
    : parseBabel(code, { sourceType: 'module' })
  const ast = babelFileNode.program
  const s = new MagicStringAST(code, { offset: -babelFileNode.start! })

  const analyzed = analyze(ast)
  transform({ ...analyzed, code, ast, s })

  return generate(s, options)
}

function analyze(ast: BabelProgram) {
  const { scope, firstNodeAfterImportDeclaration } = analyzeAst(ast)

  // keep vapor imports (`import { ... } from 'vue/vapor'`) to generate
  const vaporImports = new Set<string>()

  /**
   * collect variables that are declared with `let` to be used as `ref()`
   */

  const refVariables = [...scope.variables.values()].filter(variable => {
    return (
      variable.definition.node.type === 'VariableDeclarator' &&
      variable.declaration.type === 'VariableDeclaration' &&
      variable.declaration.kind === 'let' &&
      variable.export == undefined
    )
  })

  if (refVariables.length > 0) {
    vaporImports.add('ref')
  }

  /**
   * collect export variables (`export let ...`) to be used as props
   */

  const removableExportDeclarations = new Set<BabelNode>()
  const exportVariables = getExportVariables(scope, variable => {
    removableExportDeclarations.add(variable.export)
    removableExportDeclarations.add(variable.declaration)
  })

  /**
   * collect effectable references and label statements
   */

  // keep effectable references to use as `computed()`
  const effectables = new Set<string>()
  // keep label statements to replace with `computed`, `watchEffect` and etc
  const transformableEffectLabels = new Set<[BabelLabeledStatement, BabelIdentifier | undefined]>()

  function hasReactivityVariable(name: string): boolean {
    for (const variable of refVariables) {
      if (variable.name === name) {
        return true
      }
    }
    for (const variable of [...exportVariables.readable, ...exportVariables.writable]) {
      if (variable.name === name) {
        return true
      }
    }

    if (effectables.has(name)) {
      return true
    }

    return false
  }

  function collectRecusiveEffectables(node: BabelNode): BabelIdentifier | undefined {
    switch (node.type) {
      case 'Identifier': {
        if (!hasReactivityVariable(node.name)) {
          effectables.add(node.name)
          vaporImports.add('computed')
          return node
        }
        break
      }
      // TODO: spread pattern
      // case 'ObjectPattern': {
      //   for (const property of node.properties) {
      //     if (property.type === 'ObjectProperty') {
      //       collectRecusiveEffectables(property.value)
      //     }
      //   }
      //   break
      // }
      // case 'ArrayPattern': {
      //   for (const element of node.elements) {
      //     if (element) {
      //       collectRecusiveEffectables(element)
      //     }
      //   }
      //   break
      // }
      default: {
        return undefined
      }
    }
  }

  for (const labelStmt of scope.labels.filter(node => node.label.name === '$')) {
    if (labelStmt.body.type === 'BlockStatement') {
      // `$: { ... }`
      // TODO: strictly collect block statement
      transformableEffectLabels.add([labelStmt, undefined])
      vaporImports.add('watchEffect')
    } else if (labelStmt.body.type === 'ExpressionStatement') {
      // `$: x = 1`, `$: setCount(count + 1)` and etc
      const expression = labelStmt.body.expression
      switch (expression.type) {
        case 'CallExpression': {
          for (const arg of expression.arguments) {
            if (arg.type === 'Identifier' && hasReactivityVariable(arg.name)) {
              transformableEffectLabels.add([labelStmt, undefined])
              vaporImports.add('watchEffect')
              break
            }
          }
          break
        }
        case 'AssignmentExpression': {
          if (expression.operator === '=') {
            const id = collectRecusiveEffectables(expression.left)
            if (expression.right.type === 'Identifier') {
              if (hasReactivityVariable(expression.right.name)) {
                transformableEffectLabels.add([labelStmt, id])
              }
            } else if (expression.right.type === 'CallExpression') {
              for (const arg of expression.right.arguments) {
                if (arg.type === 'Identifier' && hasReactivityVariable(arg.name)) {
                  transformableEffectLabels.add([labelStmt, id])
                  break
                }
              }
            }
          }
          break
        }
        // No default
      }
    }
  }

  /**
   * collect store imports
   */

  const storeImportBindings = new Map<string, ImportBinding>()
  const storeImports = new Map<BabelImportDeclaration, string>()
  for (const node of ast.body) {
    switch (node.type) {
      case 'ImportDeclaration': {
        if (node.source.value === 'svelte') {
          storeImports.set(node, 'svelte')
        } else if (node.source.value === 'svelte/store') {
          storeImports.set(node, 'svelte/store')
          const imports: Record<string, ImportBinding> = {}
          walkImportDeclaration(imports, node)
          for (const [name, binding] of Object.entries(imports)) {
            if (!storeImportBindings.has(name)) {
              storeImportBindings.set(name, binding)
            }
          }
        } else {
          // TODO: svelte/animate, svelte/transition and etc
        }
        break
      }
    }
  }
  const storeVariables = getVaporStoreVariables(scope, Object.fromEntries(storeImportBindings))
  const storeReferences = getVaporStoreConvertableVariables(scope, storeVariables)

  return {
    scope,
    firstNodeAfterImportDeclaration,
    vaporImports,
    refVariables,
    removableExportDeclarations,
    exportVariables,
    effectables,
    transformableEffectLabels,
    storeImports,
    storeVariables,
    storeReferences
  }
}

type TransformContext = ReturnType<typeof analyze> & {
  code: string
  ast: BabelProgram
  s: MagicStringAST
}

function transform(context: TransformContext): void {
  transformReactivity(context)
  transformProps(context)
  transformStore(context)
  transformEffect(context)
  prependVaporImports(context)
}

function transformReactivity({ s, refVariables, code }: TransformContext): void {
  /**
   * transform `let` define variables and reference variables
   */
  for (const variable of refVariables) {
    // transform to `ref()`
    if (
      variable.export == undefined &&
      variable.definition.node.type === 'VariableDeclarator' &&
      variable.definition.node.init != undefined &&
      validateNodeTypeForVaporRef(variable.definition.node.init)
    ) {
      s.overwriteNode(
        variable.declaration,
        `const ${variable.name} = ref(${sliceCode(code, variable.definition.node.init, s.offset)})`
      )
    } else {
      continue
    }

    // transform to `.value`
    const refs = getReferences(variable)
    for (const ref of refs) {
      s.overwriteNode(ref, `${ref.name}.value`)
    }
  }
}

function sliceCode(code: string, node: BabelNode, offset?: number): string {
  return offset == undefined
    ? code.slice(node.start!, node.end!)
    : code.slice(node.start! + offset, node.end! + offset)
}

function transformEffect(context: TransformContext): void {
  const { s, transformableEffectLabels } = context
  /**
   * transform effect label statements
   */
  for (const [labelStmt, id] of transformableEffectLabels) {
    if (labelStmt.body.type === 'BlockStatement') {
      // `$: { ... }`
      s.overwriteNode(labelStmt, `watchEffect(() => ${s.sliceNode(labelStmt.body)})\n`)
    } else if (labelStmt.body.type === 'ExpressionStatement') {
      // `$: x = y`, `$: setCount(count + 1)` and etc
      const expression = labelStmt.body.expression
      switch (expression.type) {
        case 'CallExpression': {
          s.overwriteNode(labelStmt, `watchEffect(() => ${s.sliceNode(labelStmt.body)})\n`)
          break
        }
        case 'AssignmentExpression': {
          if (
            expression.operator === '=' &&
            (expression.right.type === 'Identifier' ||
              expression.right.type === 'CallExpression') &&
            id
          ) {
            s.overwriteNode(
              labelStmt,
              `const ${id.name} = computed(() => ${s.sliceNode(expression.right)})\n`
            )
            replaceLabelEffectableReferences(context.scope, id, context)
          }
          break
        }
        // No default
      }
    }
  }
}

function replaceLabelEffectableReferences(
  scope: Scope,
  id: BabelIdentifier,
  context: TransformContext
): void {
  const refs = scope.references.filter(ref => ref.start! > id.start! && ref.name === id.name)
  refs.forEach(ref => context.s.overwriteNode(ref, `${context.s.sliceNode(ref)}.value`))
  scope.children.forEach(child => replaceLabelEffectableReferences(child, id, context))
}

function transformProps(context: TransformContext): void {
  const {
    s,
    code,
    removableExportDeclarations,
    exportVariables: { readable: exportReadableVariables, writable: exportWritableVariables },
    firstNodeAfterImportDeclaration
  } = context

  /**
   * adjust order by definition node start position
   */
  for (const variables of [exportWritableVariables, exportReadableVariables]) {
    variables.sort((a, b) => a.definition.node.start! - b.definition.node.start!)
  }

  /**
   * generate code
   */
  const codes: string[] = []
  if (exportWritableVariables.length > 0) {
    for (const variable of exportWritableVariables) {
      let defaultValue = ''
      if (variable.definition.node.type === 'VariableDeclarator' && variable.definition.node.init) {
        defaultValue = sliceCode(code, variable.definition.node.init)
      }
      codes.push(
        `const ${variable.name} = defineModel('${variable.name}'${defaultValue ? `, { default: ${defaultValue} }` : ''})`
      )
    }
  }
  if (exportReadableVariables.length > 0) {
    // eslint-disable-next-line unicorn/no-array-reduce
    const exportReadableProps = exportReadableVariables.reduce((acc, variable) => {
      return variable.definition.node.type === 'VariableDeclarator' && variable.definition.node.init
        ? [...acc, `${variable.name} = ${sliceCode(code, variable.definition.node.init)}`]
        : [...acc, variable.name]
    }, [] as string[])
    codes.push(
      `const { ${exportReadableProps.join(
        ', '
      )} } = defineProps([${exportReadableVariables.map(variable => `'${variable.name}'`).join(', ')}])`
    )
  }

  /**
   * remove unnecessary variables
   */
  removableExportDeclarations.forEach(declaration => s.removeNode(declaration))

  /**
   * insert code to the first node after import declaration
   */
  if (firstNodeAfterImportDeclaration) {
    s.prependRight(firstNodeAfterImportDeclaration.start! + s.offset, codes.join('\n'))
  }
}

function transformStore(context: TransformContext): void {
  const { s, storeImports, storeVariables, storeReferences } = context

  for (const [node, source] of storeImports) {
    if (source === 'svelte') {
      s.overwriteNode(node.source, `'svelte-vapor-runtime'`)
    } else if (source === 'svelte/store') {
      const additionalSpecifiers: string[] = []
      for (const specifier of node.specifiers) {
        if (specifier.local.name === 'writable') {
          additionalSpecifiers.push('writable', 'useWritableStore')
        } else if (specifier.local.name === 'readable') {
          additionalSpecifiers.push('readable', 'useWritableStore')
        }
      }
      insertStoreComposable(s, storeVariables)
      replaceStoreIdentifier(s, storeReferences)
      s.overwriteNode(
        node,
        `import { ${[...additionalSpecifiers].join(', ')} } from 'svelte-vapor-runtime/store'`
      )
    }
  }
}

function prependVaporImports({ s, vaporImports: imports }: TransformContext): void {
  if (imports.size > 0) {
    s.prepend(`import { ${[...imports].join(', ')} } from 'vue/vapor'\n`)
  }
}

function validateNodeTypeForVaporRef(node: BabelNode): boolean {
  // TODO: more validation, binary expression, call expression and etc ...
  return (
    node.type === 'NumericLiteral' ||
    node.type === 'StringLiteral' ||
    node.type === 'BooleanLiteral' ||
    node.type === 'NullLiteral' ||
    node.type === 'BigIntLiteral' ||
    node.type === 'DecimalLiteral' ||
    node.type === 'RegExpLiteral' ||
    node.type === 'TemplateLiteral' ||
    node.type === 'ArrayExpression' ||
    node.type === 'ObjectExpression' ||
    node.type === 'NewExpression'
  )
}

function getVaporStoreVariables(scope: Scope, imports: Record<string, ImportBinding>): Variable[] {
  const variables = [] as Variable[]
  for (const [name] of Object.entries(imports)) {
    for (const variable of scope.variables.values()) {
      const def = variable.definition.node
      if (
        variable.name !== name &&
        def.type === 'VariableDeclarator' &&
        def.init &&
        def.init.type === 'CallExpression' &&
        def.init.callee.type === 'Identifier' &&
        def.init.callee.name === name
      ) {
        variables.push(variable)
      }
    }
  }
  return variables
}

function getVaporStoreConvertableVariables(scope: Scope, variables: Variable[]): BabelIdentifier[] {
  // eslint-disable-next-line unicorn/no-array-reduce
  return variables.reduce((acc, variable) => {
    return [...acc, ...collectReferences(scope, `$${variable.name}`)]
  }, [] as BabelIdentifier[])
}

function collectReferences(scope: Scope, name: string): BabelIdentifier[] {
  const refs = scope.references.filter(ref => ref.name === name) || []
  // eslint-disable-next-line unicorn/no-array-reduce
  return scope.children.reduce(
    (acc, child) => {
      return [...acc, ...collectReferences(child, name)]
    },
    [...refs] as BabelIdentifier[]
  )
}

function insertStoreComposable(s: MagicStringAST, variables: Variable[]): void {
  function makeComposable(variable: Variable, node: BabelIdentifier): string {
    return `\nconst $${variable.name} = ${node.name === 'writable' ? 'useWritableStore' : 'useReadableStore'}(${variable.name})`
  }
  for (const variable of variables) {
    const def = variable.definition.node
    if (
      def.type === 'VariableDeclarator' &&
      def.init &&
      def.init.type === 'CallExpression' &&
      def.init.callee.type === 'Identifier'
    ) {
      s.appendRight(def.end! + s.offset, makeComposable(variable, def.init.callee))
    }
  }
}

function replaceStoreIdentifier(s: MagicStringAST, references: BabelIdentifier[]): void {
  for (const ref of references) {
    s.overwriteNode(ref, `${s.sliceNode(ref)}.value`)
  }
}

function generate(
  s: MagicStringAST,
  options: TransformSvelteScriptOptions = {}
): string | { code: string; map: GenerateMap } {
  const sourceMap = !!options.sourcemap
  const id = options.id

  if (sourceMap) {
    if (!id) {
      throw new Error('`id` is required when `sourcemap` is enabled')
    }
    const gen = generateTransform(s, id)
    if (gen == undefined) {
      throw new Error('Failed to generate source map')
    }
    return gen
  } else {
    return s.toString()
  }
}
