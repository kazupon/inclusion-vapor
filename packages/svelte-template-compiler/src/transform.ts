// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transform.ts

import { isArray } from '@vue-vapor/shared'
import { IRNodeTypes } from './ir/index.ts'
import { TransformContext, newBlock } from './transforms/index.ts'

import type { RootIRNode, RootNode } from './ir/index.ts'
import type { NodeTransform, TransformOptions } from './transforms/index.ts'

// Svelte AST -> IR
export function transform(node: RootNode, options: TransformOptions = {}): RootIRNode {
  const ir: RootIRNode = {
    type: IRNodeTypes.ROOT,
    node,
    source: node.source,
    template: [],
    component: new Set(),
    directive: new Set(),
    block: newBlock(node)
  }

  const context = new TransformContext(ir, node, options)

  transformNode(context)

  return ir
}

export function transformNode(context: TransformContext): void {
  let { node } = context

  // apply transform plugins
  const { nodeTransforms } = context.options
  const exitFns: ReturnType<NodeTransform> = []
  for (const nodeTransform of nodeTransforms) {
    const onExit = nodeTransform(node, context)
    if (onExit) {
      if (isArray(onExit)) {
        exitFns.push(...onExit)
      } else {
        exitFns.push(onExit)
      }
    }
    if (context.node) {
      // node may have been replaced
      node = context.node
    } else {
      // node was removed
      return
    }
  }

  // exit transforms
  context.node = node
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }

  if (context.node.type === IRNodeTypes.ROOT) {
    context.registerTemplate()
  }
}
