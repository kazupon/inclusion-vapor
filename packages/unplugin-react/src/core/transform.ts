// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)

import createDebug from 'debug'
import { walkAST } from 'ast-kit'
import { parse as parseBabel } from '@babel/parser'
import MagicStringStack from 'magic-string-stack'
import { compile } from 'jsx-vapor-compiler'
// import { isObject, isString } from '@vue-vapor/shared'
// import { EXPORT_HELPER_ID } from './helper'
// import { createRollupError } from './utils'

// import type { UnpluginOptions } from 'unplugin'
import type { ParserPlugin } from '@babel/parser'
import type { BabelNode, JSXElement, JSXFragment } from 'jsx-vapor-compiler'
import type { ResolvedOptions } from './types'
// import type { RawSourceMap } from 'source-map-js'

const debug = createDebug('unplugin-react-vapor:core:transform')

const RE_TS = /\.tsx$/

export function transform(
  code: string,
  filepath: string,
  _options: ResolvedOptions
): ReturnType<typeof generateTransform> {
  // const { root, isProduction } = options
  debug('transforming ...', code)
  const s = new MagicStringStack(code)

  const plugins: ParserPlugin[] = ['jsx']
  const isTS = RE_TS.test(filepath)
  if (isTS) {
    plugins.push('typescript')
  }

  const babelAst = parseBabel(s.original, {
    sourceType: 'module',
    plugins
  })

  let preambleIndex = 0
  const importSet = new Set<string>()
  const delegateEventSet = new Set<string>()
  const preambleMap = new Map<string, string>()
  function _transform() {
    const rootNodes: (JSXElement | JSXFragment)[] = []
    walkAST<BabelNode>(babelAst, {
      enter(node, parent) {
        console.log('enter', node.type, parent?.type)
        switch (node.type) {
          case 'ImportDeclaration': {
            s.overwrite(node.source.start!, node.source.end!, `"react-vapor-hooks"`)
            break
          }
          case 'JSXElement': {
            if (parent?.type === 'ReturnStatement') {
              console.log('JSXElement', s.slice(node.start!, node.end!))
              rootNodes.push(node)
              this.skip()
            }
          }
        }
      }
    })

    for (const node of rootNodes) {
      // eslint-disable-next-line prefer-const
      let { code, vaporHelpers, preamble } = compile(s.slice(node.start!, node.end!), {
        mode: 'module',
        inline: true,
        isTS,
        filename: filepath
      })
      vaporHelpers.forEach(helper => importSet.add(helper))

      preamble = preamble.replaceAll(/(?<=const )t(?=(\d))/g, `_${preambleIndex}`)
      code = code.replaceAll(/(?<== )t(?=\d)/g, `_${preambleIndex}`)
      preambleIndex++

      for (const [, key, value] of preamble.matchAll(/const (_\d+) = (_template\(.*\))/g)) {
        const result = preambleMap.get(value)
        if (result) {
          code = code.replaceAll(key, result)
        } else {
          preambleMap.set(value, key)
        }
      }

      for (const [, events] of preamble.matchAll(/_delegateEvents\((.*)\)/g)) {
        events.split(', ').forEach(event => delegateEventSet.add(event))
      }

      s.overwrite(node.start!, node.end!, code)
    }
  }
  _transform()

  if (delegateEventSet.size > 0) {
    s.prepend(`_delegateEvents(${[...delegateEventSet].join(', ')});\n`)
  }

  if (preambleMap.size > 0) {
    let preambleResult = ''
    for (const [value, key] of preambleMap) {
      preambleResult += `const ${key} = ${value}\n`
    }
    s.prepend(preambleResult)
  }

  const importResult = [...importSet].map(i => `${i} as _${i}`)
  if (importResult.length > 0) {
    s.prepend(`import { ${importResult.join(', ')} } from "vue/vapor"\n`)
  }

  return generateTransform(s, filepath)
}

type SourceMap = ReturnType<typeof MagicStringStack.prototype.generateMap>

function generateTransform(
  s: MagicStringStack,
  id: string
): { code: string; map: SourceMap } | undefined {
  if (s.hasChanged()) {
    return {
      code: s.toString(),
      map: s.generateMap({
        source: id,
        includeContent: true,
        hires: 'boundary'
      })
    }
  }
}

// function isJSXElement(
//   node?: BabelNode | null,
// ): node is JSXElement | JSXFragment {
//   return !!node && (node.type === 'JSXElement' || node.type === 'JSXFragment')
// }
