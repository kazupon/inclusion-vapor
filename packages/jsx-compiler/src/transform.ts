// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import { isArray } from '@vue-vapor/shared'
import { IRNodeTypes } from './ir'
import { newBlock } from './transforms/utils'
import { TransformContext } from './transforms'

import type { RootIRNode, RootNode, BlockIRNode } from './ir'
import type { NodeTransform, TransformOptions } from './transforms'

// AST (Babel) -> IR
export function transform(node: RootNode, options: TransformOptions = {}): RootIRNode {
  const ir: RootIRNode = {
    type: IRNodeTypes.ROOT,
    node,
    source: node.source,
    template: [],
    component: new Set(),
    directive: new Set(),
    block: newBlock(node)
  }

  const context = new TransformContext(ir, node, options)

  transformNode(context)

  return ir
}

export function transformNode(context: TransformContext<BlockIRNode['node']>): void {
  let { node } = context

  // apply transform plugins
  const { nodeTransforms } = context.options
  const exitFns: ReturnType<NodeTransform> = []
  for (const nodeTransform of nodeTransforms) {
    const onExit = nodeTransform(node, context)
    if (onExit) {
      if (isArray(onExit)) {
        exitFns.push(...onExit)
      } else {
        exitFns.push(onExit)
      }
    }
    if (context.node) {
      // node may have been replaced
      node = context.node
    } else {
      // node was removed
      return
    }
  }

  // exit transforms
  context.node = node
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }

  if (context.node.type === IRNodeTypes.ROOT) {
    context.registerTemplate()
  }
}
