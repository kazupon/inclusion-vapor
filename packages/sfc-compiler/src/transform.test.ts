import { describe, test, expect } from 'vitest'
import { transformSvelteScript, transformSvelteVapor } from './transform'
import { parse } from 'svelte/compiler'
import { MagicStringAST } from 'magic-string-ast'

const jsCode = `let count = 0
const increment = () => {
  count += 1
}`

const svelteCode = `
<script>
${jsCode}
</script>

<button on:click={increment}>
  count is {count}
</button>
`

describe.skip('transformSvelteVapor', () => {
  test('MVP: counter case', () => {
    const { code } = transformSvelteVapor(svelteCode)
    expect(code).toMatchSnapshot()
  })
})

describe('transformSvelteScript', () => {
  describe('MVP: counter case', () => {
    test('code only', () => {
      const code = transformSvelteScript(jsCode)
      expect(code).toMatchSnapshot()
      expect(code).contains(`import { ref } from 'vue/vapor'`)
      expect(code).contains('let count = ref(0)')
      expect(code).contains('count.value += 1')
    })

    test('with ast', () => {
      const ast = parse(svelteCode)

      const babelFileNode = ast.instance!.content
      const svelteStr = new MagicStringAST(svelteCode)

      const { code, map } = transformSvelteScript(svelteStr.sliceNode(babelFileNode), {
        ast: ast.instance,
        id: 'test.svelte',
        sourceMap: true
      }) as { code: string; map: ReturnType<MagicStringAST['generateMap']> }
      expect(code).toMatchSnapshot()
      expect(code).contains(`import { ref } from 'vue/vapor'`)
      expect(code).contains('let count = ref(0)')
      expect(code).contains('count.value += 1')
      expect(map).toMatchSnapshot()
    })
  })
})
