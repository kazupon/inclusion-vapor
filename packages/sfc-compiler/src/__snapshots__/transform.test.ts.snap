// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformSvelteScript > MVP: counter case > code only 1`] = `
"import { ref } from 'vue/vapor'
let count = ref(0)
const increment = () => {
  count.value += 1
}"
`;

exports[`transformSvelteScript > MVP: counter case > with ast 1`] = `
"import { ref } from 'vue/vapor'

let count = ref(0)
const increment = () => {
  count.value += 1
}
"
`;

exports[`transformSvelteScript > MVP: counter case > with ast 2`] = `
SourceMap {
  "file": undefined,
  "mappings": ";AAAA;AACA,GAAG,CAAC,KAAK,CAAC,CAAC,CAAC,MAAC;AACb,KAAK,CAAC,SAAS,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACzB,CAAC,CAAC,WAAK,CAAC,CAAC,CAAC,CAAC,CAAC;AACZ,CAAC;",
  "names": [],
  "sources": [
    "test.svelte",
  ],
  "sourcesContent": [
    "
let count = 0
const increment = () => {
  count += 1
}
",
  ],
  "version": 3,
}
`;
